Std/Data/List/Perm.lean

```diff
@@ -0,0 +1,908 @@
+  ⟨fun p => (perm_append_left_iff _).1 <| perm_append_comm.trans <| p.trans perm_append_comm,
+    Perm.append_right _⟩
+
+theorem subperm_cons (a : α) {l₁ l₂ : List α} : a :: l₁ <+~ a :: l₂ ↔ l₁ <+~ l₂ :=
+  ⟨fun ⟨l, p, s⟩ => by
+    match s with
+    | .cons _ s' => exact (p.subperm_left.2 <| (sublist_cons _ _).subperm).trans s'.subperm
+    | .cons₂ _ s' => exact ⟨_, p.cons_inv, s'⟩
+  , fun ⟨l, p, s⟩ => ⟨a :: l, p.cons a, s.cons₂ _⟩⟩
+
+theorem cons_subperm_of_mem {a : α} {l₁ l₂ : List α} (d₁ : Nodup l₁) (h₁ : a ∉ l₁) (h₂ : a ∈ l₂)
+    (s : l₁ <+~ l₂) : a :: l₁ <+~ l₂ :=
+  by
+  rcases s with ⟨l, p, s⟩
+  induction s generalizing l₁
+  case slnil => cases h₂
+  case cons r₁ r₂ b s' ih =>
+    simp at h₂
+    match h₂ with
+    | .inl e =>
+      subst_vars
+      exact ⟨_ :: r₁, p.cons _, s'.cons₂ _⟩
+    | .inr m =>
+      rcases ih d₁ h₁ m p with ⟨t, p', s'⟩
+      exact ⟨t, p', s'.cons _⟩
+  case cons₂ r₁ r₂ b _ ih =>
+    have bm : b ∈ l₁ := p.subset <| mem_cons_self _ _
+    have am : a ∈ r₂ := by
+      simp only [find?, mem_cons] at h₂
+      exact h₂.resolve_left fun e => h₁ <| e.symm ▸ bm
+    rcases append_of_mem bm with ⟨t₁, t₂, rfl⟩
+    have st : t₁ ++ t₂ <+ t₁ ++ b :: t₂ := by simp
+    rcases ih (d₁.sublist st) (mt (fun x => st.subset x) h₁) am
+        (Perm.cons_inv <| p.trans perm_middle) with
+      ⟨t, p', s'⟩
+    exact
+      ⟨b :: t, (p'.cons b).trans <| (swap _ _ _).trans (perm_middle.symm.cons a), s'.cons₂ _⟩
+
+theorem subperm_append_left {l₁ l₂ : List α} : ∀ l, l ++ l₁ <+~ l ++ l₂ ↔ l₁ <+~ l₂
+  | [] => Iff.rfl
+  | a :: l => (subperm_cons a).trans (subperm_append_left l)
+
+theorem subperm_append_right {l₁ l₂ : List α} (l) : l₁ ++ l <+~ l₂ ++ l ↔ l₁ <+~ l₂ :=
+  (perm_append_comm.subperm_left.trans perm_append_comm.subperm_right).trans (subperm_append_left l)
+
+theorem Subperm.exists_of_length_lt {l₁ l₂ : List α} :
+    l₁ <+~ l₂ → length l₁ < length l₂ → ∃ a, a :: l₁ <+~ l₂
+  | ⟨l, p, s⟩, h => by
+    suffices length l < length l₂ → ∃ a : α, a :: l <+~ l₂ from
+      (this <| p.symm.length_eq ▸ h).imp fun a => (p.cons a).subperm_right.1
+    clear h p l₁
+    induction s with
+    | slnil => intro h; cases h
+    | cons a s IH =>
+      intro h
+      cases Nat.lt_or_eq_of_le (Nat.le_of_lt_succ h)
+      next h => exact (IH h).imp fun a s => s.trans (sublist_cons _ _).subperm
+      next h => exact ⟨a, s.eq_of_length h ▸ Subperm.refl _⟩
+    | cons₂ b _ IH =>
+      intro h
+      exact (IH <| Nat.lt_of_succ_lt_succ h).imp fun a s =>
+          (swap _ _ _).subperm_right.1 <| (subperm_cons _).2 s
+
+protected theorem Nodup.subperm (d : Nodup l₁) (H : l₁ ⊆ l₂) : l₁ <+~ l₂ :=
+  by
+  induction d with
+  | nil => exact ⟨nil, Perm.nil, nil_sublist _⟩
+  | cons h d IH =>
+    have ⟨H₁, H₂⟩ := forall_mem_cons.1 H
+    have := fun contra => h _ contra rfl
+    exact cons_subperm_of_mem d this H₁ (IH H₂)
+
+theorem perm_ext {l₁ l₂ : List α} (d₁ : Nodup l₁) (d₂ : Nodup l₂) :
```

Comment:
Afaik ext lemmas are usually supposed to be of the form (extensional version -> equivalence), so in this case `(∀ a, a ∈ l₁ ↔ a ∈ l₂) → l₁ ~ l₂`. Other than that though, `ext` seems to be convention for lemmas of this form.

Reply:
It seems like it's meant for simplifying perm in the case where `Nodup l1` and `Nodup l2`, rather than being used as an ext lemma.

I'd prefer a name like `Perm.subsets` or `Perm.subsets_of_nodup`. I'll look into where it's used in Mathlib.

Reply:
It's used 7 times in mathlib. 5/7 uses are to construct an iff, 2/7 are rewriting.

Reply:
I think the usual name would be `ext_iff` rather than `ext` here. Maybe `perm_ext_iff_of_nodup`?